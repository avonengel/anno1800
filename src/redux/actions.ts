import {createStandardAction} from "typesafe-actions";
import {RootState} from "./store";

export const uploadState = createStandardAction('UPLOAD_STATE')<RootState>();