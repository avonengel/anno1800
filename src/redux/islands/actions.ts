import {ADD_ISLAND, DELETE_ISLAND, IslandActionTypes, UPDATE_HOUSES, UPDATE_POPULATION} from "./types";
import {createCustomAction, createStandardAction} from "typesafe-actions";

export function createIsland(name: string): IslandActionTypes {
    return {
        type: ADD_ISLAND,
        name: name
    }
}

export function deleteIsland(id: number): IslandActionTypes {
    return {
        type: DELETE_ISLAND,
        id: id
    }
}

export const selectIsland = createStandardAction('islands/SELECT')<number>();

export function updateHouseCount(id: number, level: string, houses: number): IslandActionTypes {
    return {
        type: UPDATE_HOUSES,
        islandId: id,
        level,
        houses,
    }
}

export function updatePopulation(id: number, level: string, population: number): IslandActionTypes {
    return {
        type: UPDATE_POPULATION,
        islandId: id,
        level,
        population,
    }
}

export const renameIsland = createCustomAction('islands/RENAME',
    type => (islandId: number, name: string) => ({
        type,
        payload: {islandId, name}
    }));

export const selectPreviousIsland = createStandardAction('islands/SELECT_PREVIOUS')();
export const selectNextIsland = createStandardAction('islands/SELECT_NEXT')();
