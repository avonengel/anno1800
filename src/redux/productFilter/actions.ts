import {createStandardAction} from "typesafe-actions";

export const updateFilterType = createStandardAction('productFilter/UPDATE_FILTER_TYPE')<number>();

export const updateFilter = createStandardAction('productFilter/UPDATE_FILTER')<string>();
