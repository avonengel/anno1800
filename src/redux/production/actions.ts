import {createCustomAction} from "typesafe-actions";

// export function updateFactoryCount(islandId: number, factoryId: number, count: number): ProductionActionTypes {
export const updateFactoryCount = createCustomAction('factories/UPDATE_COUNT', type =>
    (islandId: number, factoryId: number, count: number) => ({
        type,
        payload: {islandId, factoryId, count}
    })
);

export const updateFactoryProductivity = createCustomAction('factories/UPDATE_PRODUCTIVITY', type =>
    (islandId: number, factoryId: number, productivity: number) => ({
        type, payload: {islandId, factoryId, productivity}
    })
);

export const FactoryActions = [updateFactoryCount, updateFactoryProductivity];
//(islandId: number, factoryId: number, productivity: number) => action(UPDATE_FACTORY_PRODUCTIVITY, {islandId, factoryId, productivity});
