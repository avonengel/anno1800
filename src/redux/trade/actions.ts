import {createCustomAction} from "typesafe-actions";
import {ADD_TRADE, UPDATE_TRADE_ISLANDS, UPDATE_TRADE_PRODUCT} from "./constants";

export const addTrade = createCustomAction(ADD_TRADE, type => () => ({type}));

export const updateTradeIslands = createCustomAction(UPDATE_TRADE_ISLANDS,
    type => (tradeId: number, islandId1: number, islandId2: number) => ({
        type,
        payload: {tradeId, islandId1, islandId2}
    }));
export const updateTradeProduct = createCustomAction(UPDATE_TRADE_PRODUCT,
    type => (tradeId: number, productId: number) => ({
        type,
        payload: {tradeId, productId}
    }));
