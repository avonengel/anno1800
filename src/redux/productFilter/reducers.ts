import {ActionType, createReducer} from "typesafe-actions";
import {initialState, RootState} from "../root-state";
import {updateFilter, updateFilterType} from "./actions";
import iassign from "immutable-assign";
import {FILTERS} from "../../data/filterAssets";

export const productFilterReducer = createReducer(initialState)
    .handleAction(updateFilterType, (state: RootState, action: ActionType<typeof updateFilterType>) => {
        return iassign(state, (s: RootState) => s.productFilter, (filterConfig: {filterType: number, filter: string}) => {
            const selectedFilter = FILTERS.find(f => f.guid === action.payload);
            if (selectedFilter) {
                let firstFilter: string = "";
                for (let filter in selectedFilter.options) {
                    // TODO provide helper function via Filter class?
                    firstFilter = filter;
                    break;
                }
                return {
                    ...filterConfig,
                    filterType: action.payload,
                    filter: firstFilter,
                };
            }
            return {
                ...filterConfig,
                filterType: action.payload,
            }
        });
    })
    .handleAction(updateFilter, (state, action) => {
        return iassign(state, s => s.productFilter.filter, () => action.payload);
    });