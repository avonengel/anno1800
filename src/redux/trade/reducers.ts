import {isActionOf} from "typesafe-actions";
import {addTrade, deleteTrade, updateTonsPerMinute, updateTradeIslands, updateTradeProduct} from "./actions";
import {AnyAction} from "redux";
import {RootState} from "../store";
import {deleteIsland} from "../islands/actions";

export interface Trade {
    fromIslandId: number,
    toIslandId: number,
    productId: number,
    tonsPerMinute: number,
}

export type TradeState = {
    tradesById: { [tradeId: number]: Trade };
    allTradeIds: number[];
};

export const initialTradeState: TradeState = {
    tradesById: {},
    allTradeIds: [],
};

export function tradeReducer(state: RootState, action: AnyAction) {
    if (isActionOf(deleteIsland, action)) {
        const deletedIslandId = action.payload;
        const tradesToDelete = state.trades.allTradeIds.filter(tradeId => {
            const trade = state.trades.tradesById[tradeId];
            return trade.toIslandId === deletedIslandId || trade.fromIslandId === deletedIslandId;
        });
        const tradesById: { [tradeId: number]: Trade } = {};
        const remainingTradeIds = state.trades.allTradeIds.filter(tradeId => !tradesToDelete.includes(tradeId));
        for (let tradeId of remainingTradeIds) {
            tradesById[tradeId] = state.trades.tradesById[tradeId];
        }
        let intermediate: RootState = state;
        tradesToDelete.forEach(tradeId => {
            const trade = state.trades.tradesById[tradeId];
            if (trade.toIslandId === deletedIslandId) {
                const {[tradeId]: _, ...exports} = intermediate.products[trade.fromIslandId][trade.productId].exports;
                intermediate = {
                    ...intermediate,
                    products: {
                        ...intermediate.products,
                        [trade.fromIslandId]: {
                            ...intermediate.products[trade.fromIslandId],
                            [trade.productId]: {
                                ...intermediate.products[trade.fromIslandId][trade.productId],
                                exports
                            }
                        }
                    }
                }
            } else {
                const {[tradeId]: _, ...imports} = intermediate.products[trade.toIslandId][trade.productId].imports;
                intermediate = {
                    ...intermediate,
                    products: {
                        ...intermediate.products,
                        [trade.toIslandId]: {
                            ...intermediate.products[trade.toIslandId],
                            [trade.productId]: {
                                ...intermediate.products[trade.toIslandId][trade.productId],
                                imports
                            }
                        }
                    }
                }
            }
        });
        return {
            ...intermediate,
            trades: {
                allTradeIds: remainingTradeIds,
                tradesById,
            }
        };
    } else if (isActionOf(updateTradeIslands, action)) {
        const previousTrade = state.trades.tradesById[action.payload.tradeId];
        const trade = {
            ...previousTrade,
            fromIslandId: action.payload.fromIslandId,
            toIslandId: action.payload.toIslandId,
        };
        const products = {...state.products};
        if (previousTrade && trade.productId && previousTrade.fromIslandId !== action.payload.fromIslandId) {
            if (previousTrade.fromIslandId) {
                // delete the old export
                products[previousTrade.fromIslandId] = {...products[previousTrade.fromIslandId]};
                products[previousTrade.fromIslandId][trade.productId] = {...products[previousTrade.fromIslandId][trade.productId]};
                const {[action.payload.tradeId]: _, ...newExports} = products[previousTrade.fromIslandId][trade.productId].exports;
                products[previousTrade.fromIslandId][trade.productId].exports = newExports;
            }
            if (action.payload.fromIslandId) {
                // create the new export
                products[action.payload.fromIslandId] = {...products[action.payload.fromIslandId]};
                products[action.payload.fromIslandId][trade.productId] = {...products[action.payload.fromIslandId][trade.productId]};
                products[action.payload.fromIslandId][trade.productId].exports = {...products[action.payload.fromIslandId][trade.productId].exports};
                products[action.payload.fromIslandId][trade.productId].exports[action.payload.tradeId] = trade.tonsPerMinute;
            }
        }
        if (previousTrade && trade.productId && previousTrade.toIslandId !== action.payload.toIslandId) {
            // delete the old import
            if (!!previousTrade.toIslandId) {
                products[previousTrade.toIslandId] = {...products[previousTrade.toIslandId]};
                products[previousTrade.toIslandId][previousTrade.productId] = {...products[previousTrade.toIslandId][previousTrade.productId]};
                const {[action.payload.tradeId]: _, ...newImports} = products[previousTrade.toIslandId][previousTrade.productId].imports;
                products[previousTrade.toIslandId][previousTrade.productId].imports = newImports;
            }
            if (action.payload.toIslandId) {
                // create the new import
                products[action.payload.toIslandId] = {...products[action.payload.toIslandId]};
                const newProductState = {...products[action.payload.toIslandId][trade.productId]};
                newProductState.imports = {
                    ...newProductState.imports,
                    [action.payload.tradeId]: trade.tonsPerMinute,
                };
                products[trade.toIslandId] = {
                    ...products[trade.toIslandId],
                    [trade.productId]: newProductState,
                }
            }
        }
        return {
            ...state,
            products,
            trades: {
                ...state.trades,
                tradesById: {
                    ...state.trades.tradesById,
                    [action.payload.tradeId]: trade
                }
            }

        };
    } else if (isActionOf(updateTradeProduct, action)) {
        const previousTrade = state.trades.tradesById[action.payload.tradeId];
        const trade = {
            ...state.trades.tradesById[action.payload.tradeId],
            productId: action.payload.productId,
        };
        let products = state.products;
        if (previousTrade && trade.productId !== previousTrade.productId) {
            products = {...products};
            if (previousTrade.fromIslandId) {
                const exportIslandProducts = {...products[previousTrade.fromIslandId]};
                products[previousTrade.fromIslandId] = exportIslandProducts;
                if (previousTrade.productId) {
                    // delete the old export
                    const previousExportProductState = {...exportIslandProducts[previousTrade.productId]};
                    exportIslandProducts[previousTrade.productId] = previousExportProductState;
                    const {[action.payload.tradeId]: _, ...remainingPreviousExports} = previousExportProductState.exports;
                    previousExportProductState.exports = remainingPreviousExports;
                }

                // create the new export
                const newExportProductState = {...exportIslandProducts[trade.productId]};
                exportIslandProducts[trade.productId] = newExportProductState;
                newExportProductState.exports = {
                    ...newExportProductState.exports,
                    [action.payload.tradeId]: trade.tonsPerMinute,
                };
            }
            if (previousTrade.toIslandId) {
                const importIslandProducts = {...products[previousTrade.toIslandId]};
                products[previousTrade.toIslandId] = importIslandProducts;
                if (previousTrade.productId) {
                    // delete the old import
                    const previousImportProductState = {...importIslandProducts[previousTrade.productId]};
                    importIslandProducts[previousTrade.productId] = previousImportProductState;
                    const {[action.payload.tradeId]: _remove, ...remainingPreviousImports} = previousImportProductState.imports;
                    previousImportProductState.imports = remainingPreviousImports;
                }
                // create the new import
                const newImportProductState = {...importIslandProducts[trade.productId]};
                importIslandProducts[trade.productId] = newImportProductState;
                newImportProductState.imports = {
                    ...newImportProductState.imports,
                    [action.payload.tradeId]: trade.tonsPerMinute,
                };
            }
        }
        return {
            ...state,
            products,
            trades: {
                ...state.trades,
                tradesById: {
                    ...state.trades.tradesById,
                    [action.payload.tradeId]: trade
                }
            }
        };
    } else if (isActionOf(updateTonsPerMinute, action)) {
        const previousTrade = state.trades.tradesById[action.payload.tradeId];
        const trade = {
            ...state.trades.tradesById[action.payload.tradeId],
            tonsPerMinute: action.payload.tonsPerMinute
        };
        let products = state.products;
        if (previousTrade && previousTrade.productId) {
            products = {...products};
            // update export
            if (previousTrade.fromIslandId) {
                products[trade.fromIslandId] = {...products[trade.fromIslandId]};
                products[trade.fromIslandId][trade.productId] = {...products[trade.fromIslandId][trade.productId]};
                products[trade.fromIslandId][trade.productId].exports = {...products[trade.fromIslandId][trade.productId].exports};
                products[trade.fromIslandId][trade.productId].exports[action.payload.tradeId] = trade.tonsPerMinute;
            }
            // update import
            if (previousTrade.toIslandId) {
                products[trade.toIslandId] = {...products[trade.toIslandId]};
                products[trade.toIslandId][trade.productId] = {...products[trade.toIslandId][trade.productId]};
                products[trade.toIslandId][trade.productId].imports = {...products[trade.toIslandId][trade.productId].imports};
                products[trade.toIslandId][trade.productId].imports[action.payload.tradeId] = trade.tonsPerMinute;
            }
        }

        return {
            ...state,
            products,
            trades: {
                ...state.trades,
                tradesById: {
                    ...state.trades.tradesById,
                    [action.payload.tradeId]: trade
                }
            }
        };
    } else if (isActionOf(deleteTrade, action)) {
        const previousTrade = state.trades.tradesById[action.payload];
        const {[action.payload]: tradeToDelete, ...tradesById} = state.trades.tradesById;
        const allTradeIds = state.trades.allTradeIds.filter((tradeId) => tradeId !== action.payload);
        let products = state.products;
        if (previousTrade && previousTrade.productId) {
            products = {...products};
            // update export
            if (previousTrade.fromIslandId) {
                products[previousTrade.fromIslandId] = {...products[previousTrade.fromIslandId]};
                products[previousTrade.fromIslandId][previousTrade.productId] = {...products[previousTrade.fromIslandId][previousTrade.productId]};
                const {[action.payload]: _removed, ...remainingExports} = products[previousTrade.fromIslandId][previousTrade.productId].exports;
                products[previousTrade.fromIslandId][previousTrade.productId].exports = remainingExports;
            }
            // update import
            if (previousTrade.toIslandId) {
                products[previousTrade.toIslandId] = {...products[previousTrade.toIslandId]};
                products[previousTrade.toIslandId][previousTrade.productId] = {...products[previousTrade.toIslandId][previousTrade.productId]};
                const {[action.payload]: _removed, ...remainingImports} = products[previousTrade.toIslandId][previousTrade.productId].imports;
                products[previousTrade.toIslandId][previousTrade.productId].imports = remainingImports;
            }
        }
        return {
            ...state,
            products,
            trades: {
                tradesById,
                allTradeIds
            }
        }
    } else if (isActionOf(addTrade, action)) {
        const tradesById = {...state.trades.tradesById};
        let maxId = 0;
        for (let tradeId of state.trades.allTradeIds) {
            maxId = Math.max(maxId, tradeId);
        }
        const newTradeId = maxId + 1;
        tradesById[newTradeId] = {
            productId: 0,
            toIslandId: 0,
            fromIslandId: action.payload,
            tonsPerMinute: 0,
        };
        return {
            ...state,
            trades: {
                tradesById,
                allTradeIds: [...state.trades.allTradeIds, newTradeId]
            }
        }
    }
    return state;
}