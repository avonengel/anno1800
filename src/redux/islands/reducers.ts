import {IslandState, PopulationState, UPDATE_POPULATION, UpdatePopulationAction} from "./types";
import {AnyAction} from "redux";
import {getHouses, getPopulation, getPopulationLevelByName, POPULATION_LEVELS} from "../../data/assets"
import {initialState as initialRootState, RootState} from "../store";
import {ProductState} from "../production/types";
import {getProduction} from "../production/reducers";
import {isActionOf} from "typesafe-actions";
import {createIsland, deleteIsland, renameIsland, selectNextIsland, selectPreviousIsland, updateHouseCount} from "./actions";
import iassign from "immutable-assign";

function newPopulationStateObject() {
    return POPULATION_LEVELS.reduce((map: { [level: string]: PopulationState }, level: string) => {
        map[level] = new PopulationState(level);
        return map;
    }, {});
}

export const initialState: IslandState = {
    islandIds: [1],
    islandsById: {
        1: {
            id: 1,
            name: "Ditchwater",
            population: newPopulationStateObject(),
        },
    }
};

function getEnabledProducts(products: { [productId: number]: ProductState }, excludedFactoryId?: number) {
    if (!products) {
        return [];
    }
    const result: number[] = [];
    for (let productId in products) {
        const productState = products[productId];
        if (productState) {
            if (getProduction(productState, excludedFactoryId) > 0) {
                result.push(Number(productId));
            }
        }
    }
    return result;
}

function selectNextIslandId(islandIds: number[], currentIslandId: number) {
    const currentIndex = islandIds.findIndex(value => value === currentIslandId);
    let selectedIsland = islandIds[0];
    if (currentIndex + 1 < islandIds.length) {
        selectedIsland = islandIds[currentIndex + 1];
    }
    return selectedIsland;
}

function selectPreviousIslandId(islandIds: number[], currentIslandId: number) {
    const currentIndex = islandIds.findIndex(value => value === currentIslandId);
    let selectedIsland = islandIds[islandIds.length - 1];
    if (currentIndex - 1 >= 0) {
        selectedIsland = islandIds[currentIndex - 1];
    }
    return selectedIsland;
}

export function islandReducer(state: RootState = initialRootState, action: AnyAction): RootState {
    if (!state) {
        return initialRootState;
    }
    // if (isActionOf(updateFactoryCount, action)) {
    //     const {count, factoryId, islandId} = action.payload;
    //     const factoryOutputProduct = getFactoryById(factoryId).output;
    //     // TODO honor input.noSupplyWeightCount
    //     // TODO there may be unwanted changes to population count: set population count manually, then start to add required factories in the tool -> population count is reset based on invalid house count computation from before!
    //     // idea to solve that: add an 'Options' or 'Settings' card that contains a toggle to disable computing population from houses & available products. Might as well disable the house count in general
    //     const affectedPopulationAssets = POPULATION_LEVELS
    //         .filter(level => state.island.islandsById[islandId].population[level] !== undefined
    //             && state.island.islandsById[islandId].population[level].population > 0)
    //         .map(getPopulationLevelByName)
    //         .filter(popAsset => popAsset.inputs.some(input => input.product === factoryOutputProduct && input.supplyWeight));
    //     if (affectedPopulationAssets.length === 0) {
    //         return state;
    //     }
    //     // get enabled products, but exclude the changed factory ID
    //     const enabledProducts = getEnabledProducts(state.products[islandId], factoryId);
    //     if (count > 0) {
    //         // if factories will exist in new state, add enabled product ID locally
    //         enabledProducts.push(factoryOutputProduct);
    //     }
    //     return iassign(state, (i, context) => i.island.islandsById[context.islandId].population, populations => {
    //         affectedPopulationAssets.forEach(asset => {
    //             populations[asset.name] = {
    //                 ...populations[asset.name],
    //                 population: getPopulation(asset, populations[asset.name].houses, enabledProducts),
    //             };
    //         });
    //         return populations;
    //     }, {islandId});
    // }
    if (isActionOf(renameIsland, action)) {
        const {islandId, name} = action.payload;
        return {
            ...state,
            island: {
                ...state.island,
                islandsById: {
                    ...state.island.islandsById,
                    [islandId]: {
                        ...state.island.islandsById[islandId],
                        name
                    }
                }
            }
        };
    }
    if (isActionOf(selectNextIsland, action)) {
        const islandIds = state.island.islandIds;
        if (islandIds.length > 1) {
            const selectedIsland = selectNextIslandId(islandIds, state.selectedIsland);
            return {
                ...state,
                selectedIsland
            };
        }
    }
    if (isActionOf(selectPreviousIsland, action)) {
        const islandIds = state.island.islandIds;
        if (islandIds.length > 1) {
            let selectedIsland = selectPreviousIslandId(islandIds, state.selectedIsland);
            return {
                ...state,
                selectedIsland
            };
        }
    }
    if (isActionOf(createIsland, action)) {
        const newId = Math.max(...state.island.islandIds) + 1;
        return {
            ...state,
            island: {
                islandsById: {
                    ...state.island.islandsById,
                    [newId]: {
                        name: action.payload,
                        id: newId,
                        population: newPopulationStateObject(),
                    }
                },
                islandIds: [...state.island.islandIds, newId]
            }
        };
    }
    if (isActionOf(deleteIsland, action)) {
        if (state.island.islandIds.length > 1) {
            const selectedIsland = action.payload === state.selectedIsland ? selectPreviousIslandId(state.island.islandIds, state.selectedIsland) : state.selectedIsland;
            const {[action.payload]: _, ...islandsById} = state.island.islandsById;
            const islandIds = state.island.islandIds.filter(islandId => islandId !== action.payload);
            return {
                ...state,
                island: {
                    islandIds,
                    islandsById
                },
                selectedIsland
            }
        }
    }
    if (isActionOf(updateHouseCount, action)) {
        const {islandId, level, houses} = action;
        let population = getPopulation(getPopulationLevelByName(level), houses, getEnabledProducts(state.products[islandId]));
        return iassign(state,
            (state, context) => state.island.islandsById[islandId].population[level],
            () => new PopulationState(level, houses, population));
    }
    switch (action.type) {
        case UPDATE_POPULATION:
            const popAction = action as UpdatePopulationAction;
            const islandState = {...state.island};
            islandState.islandsById = {...islandState.islandsById};
            islandState.islandsById[popAction.islandId] = {...islandState.islandsById[popAction.islandId]};
            islandState.islandsById[popAction.islandId].population = {...islandState.islandsById[popAction.islandId].population};
            islandState.islandsById[popAction.islandId].population[popAction.level] = {
                ...islandState.islandsById[popAction.islandId].population[popAction.level],
                population: popAction.population,
                houses: getHouses(getPopulationLevelByName(popAction.level), popAction.population, getEnabledProducts(state.products[popAction.islandId]))
            };

            return {...state, island: islandState};
        default:
            return state;
    }
}
