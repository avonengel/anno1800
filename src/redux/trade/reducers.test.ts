import {tradeReducer} from './reducers';
import {initialState as initialRootState, RootState} from "../store";
import {addTrade, updateTradeIslands, updateTradeProduct} from "./actions";
import {createIsland, deleteIsland} from "../islands/actions";
import {islandReducer} from "../islands/reducers";

describe('tradeReducer', () => {
    test('create new trade, select product before other island', () => {
        // Arrange
        const initialState = {...initialRootState};
        const firstIslandId = initialState.selectedIsland;
        let state = tradeReducer(initialState, addTrade(1));
        const tradeId = state.trades.allTradeIds[0];
        state = islandReducer(state, createIsland('Other'));
        const otherIslandId = state.island.islandIds[state.island.islandIds.length - 1];
        state = tradeReducer(state, updateTradeProduct(tradeId, 1010200));

        // Act
        state = tradeReducer(state, updateTradeIslands(tradeId, firstIslandId, otherIslandId));

        // Assert
        expect(state.trades.tradesById[1]).toEqual({"fromIslandId": firstIslandId, "productId": 1010200, "toIslandId": otherIslandId, "tonsPerMinute": 0});
        expect(state.products[firstIslandId][1010200]).toBeDefined();
        expect(state.products[otherIslandId][1010200]).toBeDefined();
    });

    test('create new trade, select island, select product, inverse im/export', () => {
         // Arrange
        const initialState = {...initialRootState};
        // Add island 'Other'
        let state: RootState = islandReducer(initialState, createIsland('Other'));
        const firstIslandId = initialState.selectedIsland;
        const secondIslandId = state.island.islandIds[state.island.islandIds.length-1];
        // Add trade
        state = tradeReducer(state, addTrade(firstIslandId));
        const tradeId = state.trades.allTradeIds[0];
        // Set 'Other' island as target
        state = tradeReducer(state, updateTradeIslands(tradeId, firstIslandId, secondIslandId));
        // Set product
        state = tradeReducer(state, updateTradeProduct(tradeId, 1010200));

        // Act
        state = tradeReducer(state, updateTradeIslands(tradeId, secondIslandId, firstIslandId));

        // Assert
        // TODO
    });

    test('create new trade, inverse im/export, select product, select island', () => {
        // Arrange
        const initialState = {...initialRootState};
        // Add island 'Other'
        let state: RootState = islandReducer(initialState, createIsland('Other'));
        const firstIslandId = initialState.selectedIsland;
        const secondIslandId = state.island.islandIds[state.island.islandIds.length-1];
        // Add trade
        state = tradeReducer(state, addTrade(firstIslandId));
        const tradeId = state.trades.allTradeIds[0];
        // Inverse im/export
        state = tradeReducer(state, updateTradeIslands(tradeId, state.trades.tradesById[tradeId].toIslandId, firstIslandId));
        // Set product
        state = tradeReducer(state, updateTradeProduct(tradeId, 1010200));

        // Act
        state = tradeReducer(state, updateTradeIslands(tradeId, secondIslandId, firstIslandId));
    });

    describe('delete island', () => {
        test('should delete affected trades', () => {
            // Arrange
            const initialState = {...initialRootState};
            // Add island 'Other'
            let state: RootState = islandReducer(initialState, createIsland('Other'));
            const firstIslandId = initialState.selectedIsland;
            const secondIslandId = state.island.islandIds[state.island.islandIds.length-1];
            // Add trade
            state = tradeReducer(state, addTrade(firstIslandId));
            const tradeId = state.trades.allTradeIds[0];
            state = tradeReducer(state, updateTradeIslands(tradeId, firstIslandId, secondIslandId));
            state = tradeReducer(state, updateTradeProduct(tradeId, 1010200));

            // Act
            state = tradeReducer(state, deleteIsland(secondIslandId));

            // Assert
            expect(state.trades.allTradeIds).toHaveLength(0);
            expect(state.trades.tradesById).toStrictEqual({});
        });
    });
});
