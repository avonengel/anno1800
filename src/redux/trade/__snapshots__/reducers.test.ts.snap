// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tradeReducer create new trade, inverse im/export, select product, select island 1`] = `
Object {
  "allTradeIds": Array [
    1,
  ],
  "tradesById": Object {
    "1": Object {
      "fromIslandId": 2,
      "productId": 1010200,
      "toIslandId": 1,
      "tonsPerMinute": 0,
    },
  },
}
`;

exports[`tradeReducer create new trade, select island, select product, inverse im/export 1`] = `
Object {
  "allTradeIds": Array [
    1,
  ],
  "tradesById": Object {
    "1": Object {
      "fromIslandId": 2,
      "productId": 1010200,
      "toIslandId": 1,
      "tonsPerMinute": 0,
    },
  },
}
`;
